File: backend/src/main/kotlin/com/example/backend/services/FabriquantService.kt
package com.example.backend.services

import com.example.backend.models.Fabriquant
import com.example.backend.repositories.FabriquantRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import kotlin.jvm.optionals.getOrNull

@Service
class FabriquantService(private val fabriquantRepository: FabriquantRepository) {

    fun createFabriquant(fabriquant: Fabriquant): Fabriquant = fabriquantRepository.save(fabriquant)

    fun getAllFabriquants(): List<Fabriquant> = fabriquantRepository.findAll()

    fun getAllFabriquantsPage(pageable: Pageable): Page<Fabriquant> = fabriquantRepository.findAll(pageable)

    fun updateFabriquant(id: Int, updatedFabriquant: Fabriquant): Fabriquant {
        val existingFabriquant = fabriquantRepository.findById(id).orElseThrow { Exception("Fabriquant not found") }
        existingFabriquant.nom = updatedFabriquant.nom
        existingFabriquant.adresse = updatedFabriquant.adresse
        existingFabriquant.telephone = updatedFabriquant.telephone
        existingFabriquant.email = updatedFabriquant.email
        return fabriquantRepository.save(existingFabriquant)
    }

    fun deleteFabriquant(id: Int) {
        if (!fabriquantRepository.existsById(id)) throw Exception("Fabriquant not found")
//        fabriquantRepository.deleteById(id)
       var fabriquant =  fabriquantRepository.findById(id).getOrNull()
      fabriquant?.supprimer=1
      fabriquantRepository.save(fabriquant!!)
    }
}

File: backend_py/pharmaPython/app/services/fabriquant_service.py
# services/fabriquant_service.py
from sqlalchemy.orm import Session
from sqlalchemy import desc, asc
from typing import List, Tuple, Optional
from fastapi import HTTPException

from app.models.fabriquant import Fabriquant, FabriquantCreateSchema, FabriquantSchema, FabriquantBaseSchema
from app.repositories.fabriquant_repository import FabriquantRepository


class FabriquantService:
  def __init__(self, db: Session):
    self.db = db
    self.fabriquant_repo = FabriquantRepository(db)

  def create_fabriquant(self, f: FabriquantCreateSchema) -> FabriquantSchema:
    entity = Fabriquant(**f.model_dump())
    # entity.code = 'FOR'+str(int(self.db.query(Fabriquant).count() + 2))
    entity = self.fabriquant_repo.save(entity)
    return FabriquantSchema.model_validate(entity)

  def get_all_fabriquants(self) -> List[Fabriquant]:
    return self.db.query(Fabriquant).all()

  def get_all_fabriquants_page(self, skip: int, limit: int, sort_by: str = "id", direction: str = "DESC") -> Tuple[List[Fabriquant], int]:
    q = self.db.query(Fabriquant)
    total = q.count()
    order_col = getattr(Fabriquant, sort_by, Fabriquant.id)
    q = q.order_by(desc(order_col) if direction.upper() == "DESC" else asc(order_col))
    rows = q.offset(skip).limit(limit).all()
    return rows, total

  def update_fabriquant(self, id_: int, data: FabriquantBaseSchema) -> Optional[Fabriquant]:
    f = self.db.query(Fabriquant).filter(Fabriquant.id == id_).first()
    if not f:
      return None
    # mets à jour les champs nécessaires (ex: nom, code, email, etc.)
    for attr in ["nom", "code", "email", "telephone", "adresse"]:
      if hasattr(data, attr):
        setattr(f, attr, getattr(data, attr))
    self.db.commit()
    self.db.refresh(f)
    return f

  def delete_fabriquant(self, id_: int) -> None:
    f:Fabriquant = self.db.query(Fabriquant).filter(Fabriquant.id == id_).first()
    if not f:
      raise HTTPException(status_code=404, detail="Fabriquant non trouvé")
    f.supprimer = 1
    self.db.commit()
